src/stream/reader/logfile.rs

```diff
@@ -0,0 +1,373 @@
+use crate::client::commands::log::{CatOpts, Channel, ShowOpts};
+use crate::transfer::stream::ChannelId;
+use crate::{JobTaskCount, JobTaskId, Map};
+use byteorder::ReadBytesExt;
+use colored::{Color, Colorize};
+use std::collections::BTreeMap;
+use std::fs::File;
+use std::io::{BufReader, ErrorKind, Read, Seek, SeekFrom, Write};
+use std::path::Path;
+use tako::InstanceId;
+
+pub const BLOCK_STREAM_START: u8 = 0;
+pub const BLOCK_STREAM_CHUNK: u8 = 1;
+pub const BLOCK_STREAM_END: u8 = 2;
+
+pub struct ChunkInfo {
+    position: u64,
+    size: u32, // Currently chunk is actually limited to 128kB
+}
+
+pub struct InstanceInfo {
+    channels: [Vec<ChunkInfo>; 2],
+    finished: bool,
+}
+
+pub struct TaskInfo {
+    instances: Map<InstanceId, InstanceInfo>,
+}
+
+impl TaskInfo {
+    pub fn last_instance_id(&self) -> InstanceId {
+        *self.instances.keys().max().unwrap()
+    }
+
+    pub fn last_instance(&self) -> &InstanceInfo {
+        let last_instance_id = self.last_instance_id();
+        self.instances.get(&last_instance_id).unwrap()
+    }
+    pub fn superseded(&self) -> impl Iterator<Item = (&InstanceId, &InstanceInfo)> {
+        let last_instance_id = self.last_instance_id();
+        self.instances
+            .iter()
+            .filter(move |(instance_id, _)| **instance_id != last_instance_id)
+    }
+}
+
+pub struct LogFile {
+    file: BufReader<File>,
+    index: BTreeMap<JobTaskId, TaskInfo>,
+    start_pos: u64,
+}
+
+pub struct Summary {
+    pub n_tasks: JobTaskCount,
+    pub n_streams: u64,
+    pub n_opened: u64,
+    pub stdout_size: u64,
+    pub stderr_size: u64,
+    pub n_superseded: u64,
+    pub superseded_stdout_size: u64,
+    pub superseded_stderr_size: u64,
+}
+
+enum Block {
+    StreamStart {
+        task_id: JobTaskId,
+        instance_id: InstanceId,
+    },
+    StreamChunk {
+        task_id: JobTaskId,
+        instance_id: InstanceId,
+        channel_id: ChannelId,
+        size: u32,
+    },
+    StreamEnd {
+        task_id: JobTaskId,
+        instance_id: InstanceId,
+    },
+}
+
+impl LogFile {
+    pub fn open(path: &Path) -> anyhow::Result<Self> {
+        let mut file = BufReader::new(File::open(path)?);
+        LogFile::check_header(&mut file)?;
+        let start_pos = file.stream_position()?;
+        let index = LogFile::make_index(&mut file)?;
+        Ok(LogFile {
+            file,
+            index,
+            start_pos,
+        })
+    }
+
+    fn check_header(file: &mut BufReader<File>) -> anyhow::Result<()> {
+        let mut header = [0u8; 6];
+        file.read_exact(&mut header)?;
+        if &header != b"HQ:log" {
```

Comment:
This should be put into a constant, since it's used in two different places.

Reply:
:+1: